src/main/java/hudson/tasks/junit/JUnitStep.java

```diff
@@ -0,0 +1,185 @@
+
+            String location = step.location;
+            if (location == null) {
+                location = "**/target/surefire-reports/TEST-*.xml";
+            }
+
+            LOGGER.info("Looking for JUnit tests results in " + location + " in workspace " + workspace);
+
+            TestResult result = workspace.act(new ParseResultCallable(location, buildTime, timeOnMaster, step.keepLongStdio));
+            result.tally();
+
+            // Check for empty tests
+            if (result.isEmpty()) {
+                // most likely a configuration error in the job - e.g. false pattern to match the JUnit result files
+                throw new AbortException(Messages.JUnitResultArchiver_ResultIsEmpty());
+            }
+
+            // Add the results to the build
+            synchronized (build) {
+                TestResultAction action = build.getAction(TestResultAction.class);
+                if (action == null) {
+                    action = new TestResultAction(build, result, listener);
+                    build.addAction(action);
+
+                } else {
+                    // add to existing results
+                    result.freeze(action);
+                    action.mergeResult(result, listener);
+                    build.save();
+                }
+
+                if (action.getResult().getFailCount() > 0) {
+                    build.setResult(Result.UNSTABLE);
+                }
+            }
+
+            return new TestResultSummary(result);
+        }
+    }
+
+    @Extension
+    public static final class DescriptorImpl extends AbstractStepDescriptorImpl {
+
+        public DescriptorImpl() {
+            super(ExecutionImpl.class);
+        }
+
+        @Override
+        public String getFunctionName() {
+            return "archiveTestResults";
+        }
+
+        @Override
+        public String getDisplayName() {
+            return "Archive JUnit Test Results";
+        }
+    }
+
+
+    private static final class ParseResultCallable implements FilePath.FileCallable<TestResult> {
+        private final long buildTime;
+        private final String testResults;
+        private final long nowMaster;
+        private final boolean keepLongStdio;
+
+        private ParseResultCallable(String testResults, long buildTime, long nowMaster, boolean keepLongStdio) {
+            this.buildTime = buildTime;
+            this.testResults = testResults;
+            this.nowMaster = nowMaster;
+            this.keepLongStdio = keepLongStdio;
+        }
+
+        @Override
+        public TestResult invoke(File ws, VirtualChannel channel) throws IOException {
+            final long nowSlave = System.currentTimeMillis();
+
+            FileSet fs = Util.createFileSet(ws, testResults);
+            DirectoryScanner ds = fs.getDirectoryScanner();
+
+            TestResult result = new TestResult(buildTime + (nowSlave - nowMaster), ds, keepLongStdio);
```

Comment:
Not quite sure what the time calculations here are doing (a comment would be nice), but remember that slaves may have clocks that are skewed away from those of the master.